refactor apf tests for per handler read/write timeout

tkashem opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
In support of https://github.com/kubernetes/kubernetes/pull/114189

This PR refactors the priority and fairness unit tests so these tests can be extended to assert on the behavior with per request (handler) read/write timeout 

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:
Please go through the commits in chronological order in order to make review easier, otherwise changes in indentation makes it hard to pinpoint the changes made in the tests

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

tkashem commented:
```
$ go test -race k8s.io/apiserver/pkg/server/filters -c

$ stress -p=32 ./filters.test -test.cpu=64 -test.run=TestPriorityAndFairnessWithPanicShouldNotRejectFutureRequest
5s: 83 runs so far, 0 failures
10s: 198 runs so far, 0 failures
15s: 307 runs so far, 0 failures
20s: 428 runs so far, 0 failures
25s: 539 runs so far, 0 failures
30s: 655 runs so far, 0 failures
...
5m10s: 6983 runs so far, 0 failures
5m15s: 7098 runs so far, 0 failures
5m20s: 7208 runs so far, 0 failures
5m25s: 7319 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.cpu=64 -test.run=TestPriorityAndFairnessWithRequestTimesOutBeforeHandlerWrites
5s: 0 runs so far, 0 failures
10s: 32 runs so far, 0 failures
15s: 64 runs so far, 0 failures
20s: 84 runs so far, 0 failures
25s: 96 runs so far, 0 failures
30s: 128 runs so far, 0 failures
...
9m40s: 2882 runs so far, 0 failures
9m45s: 2912 runs so far, 0 failures
9m50s: 2933 runs so far, 0 failures
9m55s: 2958 runs so far, 0 failures
10m0s: 2982 runs so far, 0 failures
```

tkashem commented:
/assign @wojtek-t @MikeSpreitzer

jiahuif commented:
/triage accepted

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -1073,33 +1059,25 @@ func TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter(t *testing.T) {
-		}
+		// second request is expected to be rejected by APF with http.StatusTooManyRequests
+		// a) we don't expect the timeout filter to time it out with http.StatusGatewayTimeout
+		// b) we don't expect a stream reset error, since the second request
```
The timeout behavior is currently still being provided by the Kube's timeout filter (handler), right? If so, does this change make this test flaky until the timeout provider is also changed?

tkashem replied:
we recently improved the apf queue timeout 
https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L400-L419

The timeout of the second request is `5s`,  so, the second request is expected to be rejected after `1.25s`, and the timeout filter is expected to send a timeout response after `5s`

Do you think this still could be flaky? I did a `stress` on this test, and it was fine. we can increase the request timeout to `10s` to increase the delay between these two events to `7.75s`, Ideally I would like to check for `http.StatusTooManyRequests` only to since this is the behavior we are expecting here.

liggitt replied:
do we need a change to the comment above:

```
		// In the scenario tested here, there is thus a race between two goroutines to respond to
		// the second request and any of their responses is allowed by the test.
```

liggitt replied:
> I did a `stress` on this test, and it was fine

did you stress with a binary built with `-race`?

tkashem replied:
> do we need a change to the comment above:

Yes, I updated the comment

MikeSpreitzer reviewed (commented):
I left a comment inline.

MikeSpreitzer reviewed (approved):
/lgtm

dims commented:
/assign @liggitt

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -916,15 +901,17 @@ func TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter(t *testing.T) {
-				// inner handler writes header and then let the request time out.
+				// after the inner handler writes, we let the request time out.
 				w.WriteHeader(http.StatusBadRequest)
+				if _, err := w.Write([]byte("hello world")); err != nil {
```
why was this change made?

tkashem replied:
the inner handler should write to the underlying `ResponseWriter` object before the request times out, I think this is the original intention of the test, given from its name.

`Write` is more interesting to us  than `WriteHeader` in relation to the timeout filter and request deadline. 
https://github.com/kubernetes/kubernetes/blob/5837de2796d6b0c17f0d26ecf16f24ca321964dc/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L196-L202

I am trying to document these behaviors in this PR https://github.com/kubernetes/kubernetes/pull/124730 as a pre-requisite for https://github.com/kubernetes/enhancements/pull/4462

Let me know your thoughts

liggitt commented:
did splitting the test cases mess with setup costs or parallelization?

on master:
```
go test k8s.io/apiserver/pkg/server/filters -run TestPriority -race -count=1
ok  	k8s.io/apiserver/pkg/server/filters	7.116s
```

on this PR (rebased on master):
```
go test k8s.io/apiserver/pkg/server/filters -run TestPriority -race -count=1
ok  	k8s.io/apiserver/pkg/server/filters	22.555s
```

it's surprising this takes 3x as long to run the ~same tests

tkashem commented:
sharing new `stress` test results:
```
$ go test -race k8s.io/apiserver/pkg/server/filters -c
```

```
stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWithPanicShouldNotRejectFutureRequest
5s: 69 runs so far, 0 failures
10s: 433 runs so far, 0 failures
15s: 761 runs so far, 0 failures
...
6m20s: 23648 runs so far, 0 failures
6m25s: 23962 runs so far, 0 failures
6m30s: 24294 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWithRequestTimesOutBeforeHandlerWrites
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 315 runs so far, 0 failures
...
5m20s: 11922 runs so far, 0 failures
5m25s: 12093 runs so far, 0 failures
5m30s: 12268 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWithHandlerPanicsAfterRequestTimesOut
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 295 runs so far, 0 failures
...
5m50s: 13137 runs so far, 0 failures
5m55s: 13360 runs so far, 0 failures
6m0s: 13592 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWithHandlerWritesBeforeRequestTimesOut
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 313 runs so far, 0 failures
...
5m10s: 11659 runs so far, 0 failures
5m15s: 11881 runs so far, 0 failures
5m20s: 12107 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWithEnqueuedRequestTimingOut
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 304 runs so far, 0 failures
...
4m55s: 10899 runs so far, 0 failures
5m0s: 11117 runs so far, 0 failures
5m5s: 11330 runs so far, 0 failures
```

tkashem commented:
It messed up the parallelization of the tests, after moving the `t.Parallel` from out of the sub-tests and in to the parent tests, i can see similar duration for test execution

On the other hand, when we run these tests with the feature gate enabled/disabled, we can't run them in parallel anymore. That's the trade-off, but for now I have restored the current configuration  

```
(This PR)
$ go test k8s.io/apiserver/pkg/server/filters -run TestPriority -race -count=1
ok  	k8s.io/apiserver/pkg/server/filters	6.888s


(master)
$ go test k8s.io/apiserver/pkg/server/filters -run TestPriority -race -count=1
ok  	k8s.io/apiserver/pkg/server/filters	7.030s
```

liggitt commented:
> On the other hand, when we run these tests with the feature gate enabled/disabled, we can't run them in parallel anymore. That's the trade-off, but for now I have restored the current configuration

hmm... that suggests we want a different factoring, where we group by feature enablement / disablement, and run these test cases in parallel with a given feature config

tkashem commented:
If i run `stress` on the tests in parallel, i see a flake
```
stress -p=64 ./filters.test -test.run=TestPriorityAndFairnessWith
5s: 0 runs so far, 0 failures
10s: 64 runs so far, 0 failures
15s: 128 runs so far, 0 failures
20s: 146 runs so far, 0 failures
25s: 192 runs so far, 0 failures

/var/folders/0c/v7fg59_s2qn6kllppph7g1kh0000gn/T/go-stress-20240626T192030-2287032175
I0626 19:20:51.404468   64492 apf_controller.go:377] Starting API Priority and Fairness config controller
I0626 19:20:51.405067   64492 apf_controller.go:382] Running API Priority and Fairness config worker
I0626 19:20:51.405135   64492 apf_controller.go:385] Running API Priority and Fairness periodic rebalancing process
I0626 19:20:51.561692   64492 apf_controller.go:377] Starting API Priority and Fairness config controller
I0626 19:20:51.561778   64492 apf_controller.go:382] Running API Priority and Fairness config worker
I0626 19:20:51.561821   64492 apf_controller.go:385] Running API Priority and Fairness periodic rebalancing process
I0626 19:20:51.562581   64492 apf_controller.go:377] Starting API Priority and Fairness config controller
I0626 19:20:51.562610   64492 apf_controller.go:382] Running API Priority and Fairness config worker
I0626 19:20:51.562624   64492 apf_controller.go:385] Running API Priority and Fairness periodic rebalancing process
I0626 19:20:51.605652   64492 apf_controller.go:377] Starting API Priority and Fairness config controller
I0626 19:20:51.605724   64492 apf_controller.go:382] Running API Priority and Fairness config worker
I0626 19:20:51.606036   64492 apf_controller.go:377] Starting API Priority and Fairness config controller
I0626 19:20:51.606089   64492 apf_controller.go:382] Running API Priority and Fairness config worker
I0626 19:20:51.606113   64492 apf_controller.go:385] Running API Priority and Fairness periodic rebalancing process
I0626 19:20:51.605773   64492 apf_controller.go:385] Running API Priority and Fairness periodic rebalancing process
E0626 19:20:51.635342   64492 timeout.go:121] "Observed a panic" panic=<
	request handler panic'd as expected - "/request/first/panic-as-expected"
	goroutine 383 [running]:
	k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1.1()
		/Users/abukashem/go/src/k8s.io/kubernetes/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:110 +0xe4
	panic({0x105146060?, 0xc000061540?})
		/opt/homebrew/Cellar/
…
30s: 256 runs so far, 1 failures (0.39%)
35s: 320 runs so far, 1 failures (0.31%)

...
6m0s: 3614 runs so far, 1 failures (0.03%)
6m5s: 3649 runs so far, 1 failures (0.03%)
6m10s: 3712 runs so far, 1 failures (0.03%)
```

I will get to the bottom of it
/hold

tkashem committed:
apiserver: refactor, all assertions should conform

all assertions should conform so they can be configured in different
ways when a single test is run with different feature configuration

tkashem committed:
apiserver: refactor, split the uber apf test into separate tests

- no change in the behavior of the test, except each sub-test
has been moved to its own parent test.

tkashem committed:
apiserver: refactor apf test, remove the sub-test

- no change in behavior
- remove the subtest from within each parent

tkashem committed:
group the apf tests so they can be run with feature configuration

- expectations are configurable
- enable the sub-tests to run parallel within each group
- capability of specifying different expectations for each group

tkashem committed:
apiserver: improve flakes for server/filter apf tests

- use unique names for priority level and flowschema, since the
subtests run in parallel, each subtest should use a unique priorty
level name so the metrics for each subtest do not collide. (we use
the same metrics registry instance for these parallel subtests)

tkashem committed:
apiserver: improve flake in server/filters apf tests

check response headers from http.Response, not from within the
server handler (by checking the http.ResponseWriter).

explanation:
for the test, request timeout threshold is set to `5s`.
The steps when a request times out as expected:
  - APF executes the inner handler
  - the inner handler inspects the response headers (it sees the
    headers, all good) `headerMatcher.inspect(t, w, fsName, plName)`
  - the inner handler blocks forever until request
    times out `<-callerRoundTripDoneCh`
  - after `5s`, request times out, timeout handler returns
    a response to the caller

let's go through the "unexpected" timeout that causes the test to
flake when the test process is severely overloaded:
- a) due to overload, the inner handler has not yet been executed
- b) `5s` elapses, request times out, timeout handler returns a
     `http.StatusGatewayTimeout` response to the caller
- c) the inner handler starts executing, some moments after `b`
- d) it starts inspecting the APF headers in the `ResponseWriter`
     object, since `b` has already happened ` w.Header()`
     returns an empty `http.Header` causing the test to fail.

tkashem committed:
apiserver: fix the apf test handlerWritesBeforeRequestTimesOut

the inner handler should write to the underlying ResponseWriter
object before the request times out, this is the original
intention of the test, given from its name

also, Write is more interesting than WriteHeader in relation to
the timeout filter and request deadline.

tkashem commented:
I was able to find the root cause of the following flake:
```
$ stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWith
5s: 0 runs so far, 0 failures

I0627 12:37:26.914401   47146 apf_controller.go:389] Shutting down API Priority and Fairness config worker

E0627 12:37:26.914807   47146 timeout.go:140] "Post-timeout activity" logger="UnhandledError" timeElapsed="1.359083ms" method="GET" path="/request/time-out-as-expected" result=null

--- FAIL: TestPriorityAndFairnessWithRequestTimesOutBeforeHandlerWrites (5.45s)
    --- FAIL: TestPriorityAndFairnessWithRequestTimesOutBeforeHandlerWrites/PerHandlerReadWriteTimeoutEnabled=false (5.45s)
        priority-and-fairness_test.go:821: Waiting for the request: "/request/time-out-as-expected" to time out
        priority-and-fairness_test.go:795: Expected APF headers to match, but got: expected HTTP header X-Kubernetes-PF-FlowSchema-UID to have value "test-fs", but got: ""
        priority-and-fairness_test.go:828: Waiting for the inner handler of the request: "/request/time-out-as-expected" to complete
        priority-and-fairness_test.go:838: Waiting for the controller to shutdown
```

The following assertion fails:
```
        priority-and-fairness_test.go:795: Expected APF headers to match, but got: expected HTTP header X-Kubernetes-PF-FlowSchema-UID to have value "test-fs", but got: ""
```

The test does not find the expected APF headers in the `ResponseWriter` for a request that is being executed by APF. 

this is the order in which the handler chain in the test is invoked:
```
-- timeout.ServeHTTP():
       |
       | -- apfHandler.ServeHTTP():
                        |
                        |  -- innerHandler.ServeHTTP():
```
 
The inner handler  inspects the APF headers:
```
			requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				// inspect the expected APF headers
                                headerMatcher.inspect(t, w, fsName, plName) 
			})
```

for the `TestPriorityAndFairnessWithRequestTimesOutBeforeHandlerWrites` test, request timeout threshold is set to `5s`.
The steps when a request times out, as expected:
  - APF executes the inner handler
  - the inner handler inspects the response headers (it sees the headers, all good) `headerMatcher.inspect(t, w, fsName, plName)`
  - the inner handler blocks forever until request times out `<-callerRoundTripDoneCh`
  - after `5s`, request times out, timeout handler returns a response to the caller


Now, let's go through the "unexpected" timeout that causes the test to flake when the test process is severely overloaded:
- a) due to overload, the inner handler has not yet been executed
- b) `5s` elapses, request times out, timeout handler returns a `http.StatusGatewayTimeout` response to the caller
- c) the inner handler starts executing, some moments after `b`
- d) it starts inspecting the APF headers in the `ResponseWriter` object, since `b` has already happened ` w.Header()` returns an empty `http.Header` causing the test to fail. 

the inner handler sees empty headers because of this:
https://github.com/kubernetes/kubernetes/blob/991e7a8c15cbf959cd67bf92fd5e8adfd6875406/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L185-L194

Options:
- a) increase request timeout, maybe from 5s to 10s - this will decrease further the probability of flake. But the test time for the package will increase
-  b) check response headers by inspecting the `http.Response`, not from within the server handler (by checking the `http.ResponseWriter`)

tkashem commented:
sharing latest `stress` results:
```
$ go test -race k8s.io/apiserver/pkg/server/filters -c

stress -p=256 ./filters.test -test.run=TestPriorityAndFairnessWithRequestDeadline
5s: 0 runs so far, 0 failures
10s: 29 runs so far, 0 failures
15s: 259 runs so far, 0 failures
...
13m20s: 29938 runs so far, 0 failures
13m25s: 30126 runs so far, 0 failures
13m30s: 30301 runs so far, 0 failures
```

tkashem commented:
> hmm... that suggests we want a different factoring, where we group by feature enablement / disablement, and run these test cases in parallel with a given feature config

@liggitt I have taken a stab at running parallel sub-tests for each group here https://github.com/kubernetes/kubernetes/pull/122924/commits/ed430b278d244350c3deb33b8dcadeb44457746f

tkashem commented:
overall `go test` time for the package has not changed:
```
(master)

go test k8s.io/apiserver/pkg/server/filters -run TestPriorityAndFairness -race -count=1
ok  	k8s.io/apiserver/pkg/server/filters	7.020s


(This PR)
❯ go test k8s.io/apiserver/pkg/server/filters -run TestPriorityAndFairness -race -count=1
ok  	k8s.io/apiserver/pkg/server/filters	7.009s
```

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -1500,3 +1433,322 @@ func TestGetRequestWaitContext(t *testing.T) {
+						a: wantHTTPStatusGatewayTimeout,
+						b: clientMustNotTimeout{},
+					},
+					secondReqVerifier: wantBoth{
+						a: wantHTTPStatusTooManyRequests,
+						b: wantBoth{
```
Maybe instead of a binary combiner it might be better to have a combiner that holds a slice of verifier.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -1500,3 +1433,322 @@ func TestGetRequestWaitContext(t *testing.T) {
+		{
+			name: withTimeoutFilter,
+			setup: func(*testing.T) func() {
+				// TODO: disable the alpha feature gate PerHandlerReadWriteTimeout
```
Alpha? Alphas are normally disabled, right? And https://github.com/kubernetes/kubernetes/pull/114189/files#diff-71e3b98f9a6bbf5b8421e26a7ba0c079f397cd8d49abacdad943c66a4f44f03dR1288 shows it as Beta.

tkashem replied:
The KEP proposes the feature as alpha, I have updated the work in progress PR: https://github.com/kubernetes/kubernetes/pull/114189/commits/0ce20dc9cb768065288478f7b9f869b3a5ac19f0

https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4460-per-request-deadline#proposal

As far as the test is concerned, it needs to make sure that the feature is disabled (either alpha or beta) in order to make sure the the timeout handler is in effect, as it is today.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -672,451 +674,394 @@ func TestApfWithRequestDigest(t *testing.T) {
+
+		case r.URL.Path == secondRequestEnqueuedPath:
+			// we expect the concurrency to be set to 1 and so this request should never be executed.
+			t.Errorf("Expected second request to be enqueued: %q", secondRequestEnqueuedPath)
```
I would say "Expected second request to be rejected from queue rather than executed"

tkashem replied:
yes, fixed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -1500,3 +1433,322 @@ func TestGetRequestWaitContext(t *testing.T) {
+
+	// each the following parent tests represent a group, and must
+	// run serially to each other
+	groups := []struct {
```
Since there is actually only one group, this looks more complicated than necessary.
Is there an expectation of more groups in the future? If so then maybe a comment about that would be helpful. If not, then I think simplifying the test code would be in order.

tkashem replied:
Yes, we will have at least two groups, with the proposed feature enabled/disabled. I added a comment

tkashem commented:
/remove-lifecycle rotten
